src/Amplitude.php

```diff
@@ -474,49 +396,167 @@ public function getOptOut()
-                'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event'
-            );
+        if ($this->optOut) {
             return;
         }
+
+        if (!empty($this->userId)) {
+            $event->userId = $this->userId;
+        }
+
+        if (!empty($this->deviceId)) {
+            $event->deviceId = $this->deviceId;
+        }
+
+        $this->checkForApiKey();
+        $this->checkForUserIdOrDeviceId($event);
+        $ch = curl_init(static::AMPLITUDE_API_URL);
+        $this->checkCurlHandle($ch);
         $postFields = [
             'api_key' => $this->apiKey,
-            'event' => json_encode($this->event),
+            'event' => json_encode($event),
         ];
         curl_setopt($ch, \CURLOPT_POSTFIELDS, $postFields);
         // Always return instead of outputting response!
         curl_setopt($ch, \CURLOPT_RETURNTRANSFER, true);
         $response = curl_exec($ch);
-        $curlErrno = curl_errno($ch);
-        if ($curlErrno) {
-            $this->logger->critical(
-                'Curl error: ' . curl_error($ch),
-                compact('curlErrno', 'response', 'postFields')
-            );
-        } else {
-            $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
-            $this->logger->log(
-                $httpCode === 200 ? Log\LogLevel::INFO : Log\LogLevel::ERROR,
-                'Amplitude HTTP API response: ' . $response,
-                compact('httpCode', 'response', 'postFields')
-            );
-        }
+        $this->checkForCurlError($ch);
+
+        $httpCode = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
+        $this->logger->log(
+            Log\LogLevel::INFO,
+            'Amplitude HTTP API response: ' . $response,
+            compact('httpCode', 'response', 'postFields')
+        );
         curl_close($ch);
     }
+
+    protected function checkForApiKey(): void
+    {
+        if (empty($this->apiKey)) {
+            throw new \LogicException(static::EXCEPTION_MSG_NO_API_KEY);
+        }
+    }
+
+    /**
+     * @param resource $ch
+     * @throws \Exception
+     */
+    protected function checkCurlHandle($ch): void
+    {
+        if (!$ch) {
+            // Could be a number of PHP environment problems, log a critical error
+            $message = 'Call to curl_init(' . static::AMPLITUDE_API_URL . ') failed, unable to send Amplitude event';
+            $this->logger->critical($message);
+            throw new \Exception($message);
+        }
+    }
+
+    /**
+     * @param resource $ch
+     * @throws AmplitudeException
+     */
+    protected function checkForCurlError($ch): void
+    {
+        if ($curlErrno = curl_errno($ch)) {
+            $message = 'Curl error: ' . curl_error($ch);
+            $context = compact('curlErrno', 'response', 'postFields');
+            $this->logger->critical($message, $context);
+            curl_close($ch);
+            throw new \Exception($message);
```

Comment:
Same here, either need to update the readme, or keep the current philosophy of "exceptions for code related errors, log for connection problems or anything else not due to code".

Reply:
For posterity, I just realized the "same here" is referring to a comment I ended up removing...  The original comment was basically what I commented on #16 that currently the design is that it only throws exceptions for "coding errors" not "connection issues".  

While that is how it is currently designed, if changing that allows for much cleaner code in the library I think maybe we can change that, as long as we also update the docs to make it very clear you always need a try/catch around any amplitude calls to avoid possible downtime yourself.